Mathlib/Data/List/GroupBy.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2024 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+import Mathlib.Data.List.Chain
+
+/-!
+# Group a list by a relation
+
+This file provides the basic API for `List.groupBy`. The main results are the following:
+
+- `List.join_groupBy`: the lists in `List.groupBy` join to the original list.
+- `List.nil_not_mem_groupBy`: the empty list is not contained in `List.groupBy`.
+- `List.chain'_of_mem_groupBy`: any two adjacent elements in a list in `List.groupBy` are related by
+  the specified relation.
+- `List.chain'_last_head_groupBy`: the last element of each list in `List.groupBy` is not related
+  to the first element of the next list.
+-/
+
+namespace List
+
+variable {α : Type*} {m : List α}
+
+@[simp]
+theorem groupBy_nil (r : α → α → Bool) : groupBy r [] = [] :=
+  rfl
+
+private theorem groupBy_loop_eq_append {r : α → α → Bool} {l : List α} {a : α} {g : List α}
```

Comment:
I know it's private but just in case it gets unprivated later might as well give it the correct name
```suggestion
private theorem groupByLoop_eq_append {r : α → α → Bool} {l : List α} {a : α} {g : List α}
```